src/components/Table/utils/RowState.ts

```diff
@@ -126,7 +148,11 @@ export class RowState {
   private get visibleChildren(): RowState[] {
     // The keptGroup should treat all of its children as visible, as this makes select/unselect all work.
     if (this.row.kind === KEPT_GROUP) return this.children ?? [];
-    return this.children?.filter((c) => c.isMatched === true) ?? [];
+    // Ignore hard-deleted rows, i.e. from `api.deleteRows`; in theory any hard-deleted
+    // rows should be removed from `this.children` anyway, by a change to `props.rows`,
+    // but just in case the user calls _only_ `api.deleteRows`, and expects the row to
+    // go away, go ahead and filter them out here.
+    return this.children?.filter((c) => c.isMatched === true && c.removed !== "hard") ?? [];
```

Comment:
This change was necessary since `api.deleteRows` no longer actively remove rows from `parent.children`, b/c that was the one spot that needed `RowState.parent` to go up from the RowState to its parent.

Granted, introducing the notion of soft/hard deleted seems a little complicated, but I think it's worth being able to remove the `RowState.parent` pointer and stop trying to rewrite `children` collections.